### PR TITLE
Bugfix/complaint improvements

### PR DESCRIPTION
**Pull request by krusche**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

Some REST calls in the `ComplaintResource` and `ComplaintResponseResource` send too much information to the client that should be hidden instead. Also the security checks are outdated or missing.

To address these problems, I improved the security checks, removed outdated API calls and hide more information in the JSON payloads of the responses.

There are some missing TODOs when it comes to API design in this PR that we can also solve in a follow up PR.

As I changed the information the client receives in the REST Calls and that can be sent back to the server in subsequent calls, we need to check that the complaint functionality is still working correctly in all cases! Testing is needed for the following:

- [x] Create complaint / more feedback request (student)
- [x] Accept / reject complaint (tutor)
- [x] Answer more feedback request (tutor)
- [x] Display complaints / more feedback requests in the tutor / instructor course / exercise dashboards (tutor)
- [x] View addressed complaint / more feedback request (student)

**This should be tested for multiple exercise types**. In addition to functional testing, the REST responses should be inspected, if all information is hidden that students / tutors are not allowed to see, this includes at least the following:

**Students should see as few objects as possible in these network calls and must not see:**
- The identity of the tutor
- Certain elements of the exercise (e.g. grading instructions, sample solution, etc.), the best is if the exercise is just not available

**Tutors should see as few objects as possible in these network calls and must not see:**
- The identity of the student

**Ideally we also update our test cases / write new test cases to automatically test these cases!**

**Who can help with finishing the implementation?**

### CODE REVIEW

**Comment by beatrisbu** [comment]
I tested creating/rejecting/viewing complaints and requests for more feedback with the corresponding roles for text, modeling and file upload exercises and for everything works as it should. 
I also did not see any information about students/tutors/exercises in the REST responses.

**Comment by ungaralex** [comment]
Added some integration tests for the `ComplaintResource` checking if there is sensitive data in the response (depending on the role). 
We are still missing tests for `ComplaintResponseResource`

**Comment by maxr96** [comment]
I extended some existing and added integration tests to check that submitting and getting complaint response from `ComplaintResponseResource` doesn't leak any sensitive data.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java [hunk 1]
```diff
@@ -69,7 +69,6 @@ public ComplaintResponseResource(ComplaintResponseRepository complaintResponseRe
     @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
     public ResponseEntity<ComplaintResponse> createComplaintResponse(@RequestBody ComplaintResponse complaintResponse) throws URISyntaxException {
         log.debug("REST request to save ComplaintResponse: {}", complaintResponse);
-        // TODO authorizationCheckService.isAtLeastTeachingAssistantForExercise missing
```
Comment (maxr96) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResponseResource.java]: This is done in the `createComplaintResponse` method, so I removed the TODO

File: src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResource.java [hunk 2]
```diff
@@ -117,25 +120,32 @@ public ComplaintResource(AuthorizationCheckService authCheckService, ExerciseSer
...
-
-                if (authCheckService.isAtLeastTeachingAssistantForExercise(exercise)) {
-                    // filter student information if user is not instructor but at least teaching assistant (means that user is teaching assistant)
-                    complaint.get().filterSensitiveInformation();
-                    participation.filterSensitiveInformation();
-                }
-            }
+        var optionalComplaint = complaintService.getByResultId(resultId);
+        if (optionalComplaint.isEmpty()) {
+            return notFound();
```
Comment (madwau) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ComplaintResource.java]:
```suggestion
            return ResponseEntity.ok().build();
```

If there is no complaint for a result, the frontend is expecting a response with status code 200 and an empty body, not a 404 response. This behavior is also specified in the JSDoc string. The frontend is using this endpoint both to retrieve a complaint for a certain result and to check whether such a complaint exists in the first place.

Since the non-existence of a complaint for a result is not an unexpected error in this case, this should of course not be relayed to the user, which happens at the moment as @HanyaElhashemy noticed while testing. See screenshot below:

OnPaste 20191216-171636

I'm not sure which approach conforms better with REST standards when the API user knows that the resource's existence is optional. I guess a 404 response would be more explicit than an empty body but than the frontend needs to handle this "error". In this particular case, a 404 response could also be used to convey that the result for which the complaint is requested does not exist, so its meaning is a bit ambivalent here anyway.

Long story short, I think the best course of action right now is probably to simply restore the previous behavior and return an empty body if there is no complaint (see code suggestion).

**Comment by madwau** [comment]
> image
> I found another issue while testing.
> The name of the programming exercise is missing in the brackets after `Evaluate the complaint` and the `Download Repos` button is disabled.

I looked into this. The missing or broken parts of the UI are a result of the response body trimming for requests to `/api/complaints/result/{resultId}`. Below is a sample response where the red outline shows what was removed from the response by this PR and the green outlines show which attributes are needed by the dialog (`programming-assessment-manual-result-dialog`).

OnPaste 20191216-174910

**Comment by krusche** [comment]
@HanyaElhashemy @madwau thanks for the feedback. I was able to fix the problem.
I also fixed some additional bugs and improved the usability of the dialog for programming exercises.
